x-pack/plugins/event_log/server/plugin.ts

```diff
@@ -158,5 +154,20 @@ export class Plugin implements CorePlugin<IEventLogService, IEventLogClientServi
       event: { action: ACTIONS.stopping },
       message: 'eventLog stopping',
     });
+
+    this.systemLogger.info('shutdown: waiting to finish');
+    await this.esContext?.shutdown();
+    this.systemLogger.info('shutdown: finished');
```

Comment:
What is `esContext?.shutdown` doing exactly?

Just want to point out that `plugin.stop` will only be called in case of graceful shutdowns. Process termination will obviously not call this, or the process may even be killed during this invocation. So this have to be resilient to such scenarios.

Reply:
`esContext.shutdown()` will "stop" the rxjs pipeline batching event log documents, and wait till the final batch is written out (bulk indexed into the event log index).  Assuming the general rxjs "complete" processing is very quick, the only latency will be from the bulk index call, which would be max 100 docs (max ~1K in size)

Totally understand that it only gets called under "nice" circumstances.  We don't really have a better story though, and given the current buffering (100 docs or 1 sec elapsed), it's shouldn't lose too much even in an OOM kind of non-clean shutdown.

The event log is historic data, and we've always treated it as not critical data - it's not a source of truth.

And I think the story with this is better than what it's replacing, which was doing `setImmediate()`s to force the indexing of individual documents off the main loop, which could have gotten really chaotic (but we never saw that).

We're also now getting the "event log stopped" message that we've been logging since the very beginning, actually indexed - before this, it was never indexed, as presumably the Kibana process never waited for unfinished `setImmediate()` processing to finish (and rightly so!).

Still, it does _scare_ me a bit!

Reply:
I agree that awaiting for this in case of 'normal' shutdown is way better than a `setImmediate` implementation (which was likely never finishing as the delay between plugins shutdown and termination is rather quick).

Reply:
Created https://github.com/elastic/kibana/issues/83612 FYI